ext/shmop/shmop.c

```diff
@@ -66,22 +59,59 @@ zend_module_entry shmop_module_entry = {
 ZEND_GET_MODULE(shmop)
 #endif
 
-/* {{{ rsclean
- */
-static void rsclean(zend_resource *rsrc)
+zend_class_entry *shmop_ce;
+static zend_object_handlers shmop_object_handlers;
+
+static inline php_shmop *shmop_from_obj(zend_object *obj)
+{
+	return (php_shmop *)((char *)(obj) - XtOffsetOf(php_shmop, std));
+}
+
+#define Z_SHMOP_P(zv) shmop_from_obj(Z_OBJ_P(zv))
+
+static zend_object *shmop_create_object(zend_class_entry *class_type)
+{
+	php_shmop *intern = zend_object_alloc(sizeof(php_shmop), class_type);
+
+	zend_object_std_init(&intern->std, class_type);
+	object_properties_init(&intern->std, class_type);
+	intern->std.handlers = &shmop_object_handlers;
+
+	return &intern->std;
+}
+
+static zend_function *shmop_get_constructor(zend_object *object)
+{
+	zend_throw_error(NULL, "Cannot directly construct Shmop, use shmop_open() instead");
+	return NULL;
+}
+
+static void shmop_free_obj(zend_object *object)
 {
-	struct php_shmop *shmop = (struct php_shmop *)rsrc->ptr;
+	php_shmop *shmop = shmop_from_obj(object);
 
 	shmdt(shmop->addr);
-	efree(shmop);
+
+	zend_object_std_dtor(&shmop->std);
 }
-/* }}} */
 
 /* {{{ PHP_MINIT_FUNCTION
  */
 PHP_MINIT_FUNCTION(shmop)
 {
-	shm_type = zend_register_list_destructors_ex(rsclean, NULL, "shmop", module_number);
+	zend_class_entry ce;
+	INIT_CLASS_ENTRY(ce, "Shmop", class_Shmop_methods);
+	shmop_ce = zend_register_internal_class(&ce);
+	shmop_ce->ce_flags |= ZEND_ACC_FINAL;
```

Comment:
ACC_FINAL seems not necessary. I mean seems no reason to prevent user from extending this, right?